src/signature_builder.rs

```diff
@@ -0,0 +1,145 @@
+//! signature index construction as builder pattern
+
+use super::*;
+
+use std::default::Default;
+
+// marker trait for stages
+pub trait ConstructionStage {}
+
+pub struct Empty;
+pub struct WithDigest;
+pub struct WithSignature;
+
+impl ConstructionStage for Empty {}
+
+impl ConstructionStage for WithDigest {}
+
+impl ConstructionStage for WithSignature {}
+
+/// base signature header builder
+///
+/// T describes the stage and can be one of Empty, WithDigest, WithSignature
+pub struct SignatureHeaderBuilder<T>
+where
+    T: ConstructionStage,
+{
+    entries: Vec<IndexEntry<IndexSignatureTag>>,
+    // TODO right now this is always 0
+    // TODO verify what this actually is
+    // TODO and how it must be constructed
+    offset: i32,
+    phantom: std::marker::PhantomData<T>,
```

Comment:
I think the idea of a compiler driven state machine here is really cool!

Reply:
It's a good way to limit state transitions (and each builder itself is a statemachine) and avoid bug reports :)